File: pump/omnipod-eros/src/test/java/info/nightscout/androidaps/plugins/pump/omnipod/eros/driver/communication/AapsOmnipodErosManagerTest.java
package info.nightscout.androidaps.plugins.pump.omnipod.eros.driver.communication;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import org.joda.time.Duration;
import org.junit.jupiter.api.Test;

import java.util.List;

import info.nightscout.androidaps.plugins.pump.omnipod.eros.driver.definition.schedule.BasalSchedule;
import info.nightscout.androidaps.plugins.pump.omnipod.eros.driver.definition.schedule.BasalScheduleEntry;
import info.nightscout.androidaps.plugins.pump.omnipod.eros.manager.AapsOmnipodErosManager;
import info.nightscout.interfaces.profile.Profile;

public class AapsOmnipodErosManagerTest {

    @Test
    public void validProfile() {
        Profile profile = mock(Profile.class);

        when(profile.getBasalValues()).thenReturn(new Profile.ProfileValue[]{
                new Profile.ProfileValue(0, 0.5),
                new Profile.ProfileValue(18000, 1.0),
                new Profile.ProfileValue(50400, 3.05)
        });

        BasalSchedule basalSchedule = AapsOmnipodErosManager.mapProfileToBasalSchedule(profile);

        List<BasalScheduleEntry> entries = basalSchedule.getEntries();
        assertEquals(3, entries.size());

        BasalScheduleEntry entry1 = entries.get(0);
        assertEquals(Duration.standardSeconds(0), entry1.getStartTime());
        assertEquals(0.5D, entry1.getRate(), 0.000001);

        BasalScheduleEntry entry2 = entries.get(1);
        assertEquals(Duration.standardSeconds(18000), entry2.getStartTime());
        assertEquals(1.0D, entry2.getRate(), 0.000001);

        BasalScheduleEntry entry3 = entries.get(2);
        assertEquals(Duration.standardSeconds(50400), entry3.getStartTime());
        assertEquals(3.05D, entry3.getRate(), 0.000001);
    }

    @Test
    public void invalidProfileNullProfile() {
        assertThrows("Profile can not be null", IllegalArgumentException.class, () -> AapsOmnipodErosManager.mapProfileToBasalSchedule(null));
    }

    @Test
    public void invalidProfileNullEntries() {
        assertThrows("Basal values can not be null", IllegalArgumentException.class, () -> AapsOmnipodErosManager.mapProfileToBasalSchedule(mock(Profile.class)));
    }

    @Test
    public void invalidProfileZeroEntries() {
        Profile profile = mock(Profile.class);

        when(profile.getBasalValues()).thenReturn(new Profile.ProfileValue[0]);

        assertThrows("Entries can not be empty", IllegalArgumentException.class, () -> AapsOmnipodErosManager.mapProfileToBasalSchedule(profile));
    }

    @Test
    public void invalidProfileNonZeroOffset() {
        Profile profile = mock(Profile.class);

        when(profile.getBasalValues()).thenReturn(new Profile.ProfileValue[]{
                new Profile.ProfileValue(1800, 0.5)
        });

        assertThrows("First basal schedule entry should have 0 offset", IllegalArgumentException.class, () -> AapsOmnipodErosManager.mapProfileToBasalSchedule(profile));
    }

    @Test
    public void invalidProfileMoreThan24Hours() {
        Profile profile = mock(Profile.class);

        when(profile.getBasalValues()).thenReturn(new Profile.ProfileValue[]{
                new Profile.ProfileValue(0, 0.5),
                new Profile.ProfileValue(86400, 0.5)
        });

        assertThrows("Invalid start time", IllegalArgumentException.class, () -> AapsOmnipodErosManager.mapProfileToBasalSchedule(profile));
    }

    @Test
    public void invalidProfileNegativeOffset() {
        Profile profile = mock(Profile.class);

        when(profile.getBasalValues()).thenReturn(new Profile.ProfileValue[]{
                new Profile.ProfileValue(-1, 0.5)
        });

        assertThrows("Invalid start time", IllegalArgumentException.class, () -> AapsOmnipodErosManager.mapProfileToBasalSchedule(profile));
    }

    @Test
    public void roundsToSupportedPrecision() {
        Profile profile = mock(Profile.class);

        when(profile.getBasalValues()).thenReturn(new Profile.ProfileValue[]{
                new Profile.ProfileValue(0, 0.04),
        });

        BasalSchedule basalSchedule = AapsOmnipodErosManager.mapProfileToBasalSchedule(profile);
        BasalScheduleEntry basalScheduleEntry = basalSchedule.getEntries().get(0);

        assertEquals(0.05D, basalScheduleEntry.getRate(), 0.000001);
    }
}

File: pump/omnipod-eros/src/test/java/info/nightscout/androidaps/plugins/pump/omnipod/eros/driver/communication/AapsOmnipodErosManagerTest.kt
package info.nightscout.androidaps.plugins.pump.omnipod.eros.driver.communication

import app.aaps.core.interfaces.profile.Profile
import app.aaps.core.interfaces.profile.Profile.ProfileValue
import info.nightscout.androidaps.plugins.pump.omnipod.eros.manager.AapsOmnipodErosManager
import org.joda.time.Duration
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import kotlin.test.assertFailsWith

internal class AapsOmnipodErosManagerTest {

    @Test fun validProfile() {
        val profile = Mockito.mock(Profile::class.java)
        Mockito.`when`(profile.getBasalValues()).thenReturn(
            arrayOf(
                ProfileValue(0, 0.5),
                ProfileValue(18000, 1.0),
                ProfileValue(50400, 3.05)
            )
        )
        val basalSchedule = AapsOmnipodErosManager.mapProfileToBasalSchedule(profile)
        val entries = basalSchedule.entries
        Assertions.assertEquals(3, entries.size)
        val entry1 = entries[0]
        Assertions.assertEquals(Duration.standardSeconds(0), entry1.startTime)
        Assertions.assertEquals(0.5, entry1.rate, 0.000001)
        val entry2 = entries[1]
        Assertions.assertEquals(Duration.standardSeconds(18000), entry2.startTime)
        Assertions.assertEquals(1.0, entry2.rate, 0.000001)
        val entry3 = entries[2]
        Assertions.assertEquals(Duration.standardSeconds(50400), entry3.startTime)
        Assertions.assertEquals(3.05, entry3.rate, 0.000001)
    }

    @Test fun invalidProfileNullProfile() {
        assertFailsWith<IllegalArgumentException> { AapsOmnipodErosManager.mapProfileToBasalSchedule(null) }
    }

    @Test fun invalidProfileNullEntries() {
        assertFailsWith<IllegalArgumentException> {
            AapsOmnipodErosManager.mapProfileToBasalSchedule(Mockito.mock(Profile::class.java))
        }
    }

    @Test fun invalidProfileZeroEntries() {
        val profile = Mockito.mock(Profile::class.java)
        Mockito.`when`(profile.getBasalValues()).thenReturn(emptyArray())
        assertFailsWith<IllegalArgumentException> { AapsOmnipodErosManager.mapProfileToBasalSchedule(profile) }
    }

    @Test fun invalidProfileNonZeroOffset() {
        val profile = Mockito.mock(Profile::class.java)
        Mockito.`when`(profile.getBasalValues()).thenReturn(
            arrayOf(
                ProfileValue(1800, 0.5)
            )
        )
        assertFailsWith<IllegalArgumentException> { AapsOmnipodErosManager.mapProfileToBasalSchedule(profile) }
    }

    @Test fun invalidProfileMoreThan24Hours() {
        val profile = Mockito.mock(Profile::class.java)
        Mockito.`when`(profile.getBasalValues()).thenReturn(
            arrayOf(
                ProfileValue(0, 0.5),
                ProfileValue(86400, 0.5)
            )
        )
        assertFailsWith<IllegalArgumentException> { AapsOmnipodErosManager.mapProfileToBasalSchedule(profile) }
    }

    @Test fun invalidProfileNegativeOffset() {
        val profile = Mockito.mock(Profile::class.java)
        Mockito.`when`(profile.getBasalValues()).thenReturn(
            arrayOf(
                ProfileValue(-1, 0.5)
            )
        )
        assertFailsWith<IllegalArgumentException> { AapsOmnipodErosManager.mapProfileToBasalSchedule(profile) }
    }

    @Test fun roundsToSupportedPrecision() {
        val profile = Mockito.mock(Profile::class.java)
        Mockito.`when`(profile.getBasalValues()).thenReturn(
            arrayOf(
                ProfileValue(0, 0.04)
            )
        )
        val basalSchedule = AapsOmnipodErosManager.mapProfileToBasalSchedule(profile)
        val basalScheduleEntry = basalSchedule.entries[0]
        Assertions.assertEquals(0.05, basalScheduleEntry.rate, 0.000001)
    }
}
